### PR TITLE
Add glad library for glfw.

### PR DESCRIPTION
**Pull request by cloudwebrtc**
https://github.com/flutter/engine/pull/9822 Depends on this patch.

### CODE REVIEW

**Comment by cloudwebrtc** [comment]
@stuartmorgan Please review the dependency patch for #9822

**Comment by stuartmorgan** [comment]
I believe these are generated, rather than something we could pull with DEPS. But in that case we need a readme in the folder that clearly explains where it came from, how it's generated, how to update it, etc.

**Comment by cloudwebrtc** [comment]
@cbracken @stuartmorgan The `glad` sources is actually included in the GLFW repo, the original file is from the [third_party/glfw](https://fuchsia.googlesource.com/third_party/glfw/+/999f3556fdd80983b10051746264489f2cb1ef16/deps/) repository of fuchsia, so before I Modify BUILD.gn directly in a [PR](https://github.com/flutter/buildroot/pull/275) of [buildtools](https://github.com/flutter/buildroot/pull/275/commits/c4667a7f095d153798cc01d9ec74d6bac6597ef5) to create a target for `glad`.

I think there are two options here:

1, If we don't want a duplicate file in flutter/engine, we can modify `buildroot/build/secondary/third_party/glfw/BUILD.gn` to add a `glad` target,  because `glad.h/glad.c` files already exist.

I looked at the new version of `fuchsia/third_party/glfw` repo, which also contains the `glad` file. The latest version of glfw separates glad.h from the gl and vulkan versions. I checked the glad/gl.h The necessary function is compatible with the current version.

2, Copy the `glad` source from the `GLFW` repository, create a third-party library separately, add a separate BUILD.gn, readme, upgrade instructions. When you need to upgrade, we need to copy again from `fuchsia/third_party/glfw` repo or use https://glad.dav1d.de to generate `glad` source code.

I personally prefer the first one, create a `glad` target to `buildroot/build/secondary/third_party/glfw/BUILD.gn`, because there is no need to add any source code, and it will not increase the upgrade workload.

**Comment by stuartmorgan** [comment]
> 1, If we don't want a duplicate file in flutter/engine, we can modify `buildroot/build/secondary/third_party/glfw/BUILD.gn` to add a `glad` target, because `glad.h/glad.c` files already exist.

As I explained before, I don't this this is a good idea. From what I can see `glad` is not conceptually part of GLFW, but a totally different project. The fact that some GLFW *example* code happens to also use `glad` doesn't mean that we should use that copy; our third-party dependencies should be explicit, not included transitively from the third-party dependencies of our third-party dependencies.

> 2, Copy the `glad` source from the `GLFW` repository, create a third-party library separately, add a separate BUILD.gn, readme, upgrade instructions. When you need to upgrade, we need to copy again from `fuchsia/third_party/glfw` repo

This doesn't seem conceptually much different from the above, although at least it makes the dependencies explicit.

> use https://glad.dav1d.de to generate `glad` source code.

If that's the authoritative source, then we should do that. I thought that's what I had suggested before, and assumed is what this patch was.

> I personally prefer the first one, create a `glad` target to `buildroot/build/secondary/third_party/glfw/BUILD.gn`, because there is no need to add any source code

It's not clear to me why this is a goal.

> and it will not increase the upgrade workload.

Assuming that the version generated for use in a GLFW example app stays compatible with our needs.

**Comment by cloudwebrtc** [comment]
It seems that https://gen.glad.sh is the authoritative source, I found it in the glfw official source file and has a link with parameters.

https://gen.glad.sh/#api=gl%3Acompatibility%3D3.3&extensions=GL_ARB_multisample%2CGL_ARB_robustness%2CGL_KHR_debug&generator=c&options=

In the case where the gl version number is fixed, the generated file does not seem to change.

**Comment by cloudwebrtc** [comment]
@stuartmorgan I created a README.md, please review.
Also, I tried to create the generated source code from https://gen.glad.sh, but the gladLoadGLLoader function will be missing, which will cause compilation errors, so follow the https://fuchsia.googlesource.com/third_party/glfw/deps version upgrade possible Will be more reliable.

File: third_party/glad/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Introduction
+The `GLAD` library is used for the texture of glfw, which is used to create textures to render RGBA data.
+
+# Notice
+The current `glad` source is a copy from `fuchsia/third_party/glfw/deps` and the first version is copied from https://fuchsia.googlesource.com/third_party/glfw/+/999f3556fdd80983b10051746264489f2cb1ef16/deps.
```
Comment (stuartmorgan) [inline third_party/glad/README.md]: I'm not comfortable with checking in a file that we don't apparently understand how to create from the authoritative source, and thus don't know how to update ourselves if we needed to update it.

The [version of this file on master](https://fuchsia.googlesource.com/third_party/glfw/+/refs/heads/master/deps/glad/gl.h) contains the exact details of constructing it from the source, which seems far better. I'm not clear on why we can't do the same thing here instead of relying on this older copy of unclear origin.

Comment (cloudwebrtc) [reply third_party/glad/README.md]: Sorry, I have always thought that the glad version must be with glfw. After I tested that the glad file from http://glad.sh can still work fine, I proved that I was wrong, glfw and glad are working independently. So, you said that generating `glad` from an authoritative source is correct. I am working on the engine's texture rendering test and will update the PR later.

**Comment by cloudwebrtc** [comment]
@stuartmorgan PR updated, please review.

File: third_party/glad/BUILD.gn [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Copyright 2014 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
```
Comment (stuartmorgan) [inline third_party/glad/BUILD.gn]: Is this file copied from Chromium?

If not, it should use the standard Flutter engine copyright block.

File: third_party/glad/BUILD.gn [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+config("glad_public") {
+  include_dirs = [ "include" ]
+}
+
+source_set("glad") {
+  public_configs = [ ":glad_public" ]
+
+  defines = []
```
Comment (stuartmorgan) [inline third_party/glad/BUILD.gn]: Why is this here if it's empty?

File: third_party/glad/BUILD.gn [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
```
Comment (stuartmorgan) [inline third_party/glad/BUILD.gn]: I would expect at least `gl.h` to be in a `public`; without that you should be getting GN errors in your other patch.

File: third_party/glad/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+Step:
+
+1, Add APIs:
+
+gl : `Version 3.3`
+
+2, Select Extensions:
+* `GL_ARB_multisample`
+* `GL_ARB_robustness`
+* `GL_KHR_debug`
```
Comment (stuartmorgan) [inline third_party/glad/README.md]: To confirm: your texture patch requires all of these extensions?

File: third_party/glad/README.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# Introduction
+
+The `GLAD` library is used for the texture of glfw, which is used to create textures to render RGBA data.
+
+# Notice
```
Comment (stuartmorgan) [inline third_party/glad/README.md]: s/Notice/Updating/

File: third_party/glad/README.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+The `glad` source is generated from https://gen.gald.sh.
+
+Step:
```
Comment (stuartmorgan) [inline third_party/glad/README.md]: Nit: s/Step/Steps/

File: third_party/glad/README.md (hunk 5 above)
Comment (cloudwebrtc) [reply third_party/glad/README.md]: Well, I tested it again, no need to add any extensions, so remove it.

**Comment by cloudwebrtc** [comment]
@stuartmorgan All done,  please review again.

**Comment by stuartmorgan** [comment]
@Hixie I'm going to need to update the license script; should I make this error more generic?
```
=============================== ERROR ===============================
The total license count in flutter/ci/licenses_golden/licenses_flutter
changed from 2 to 4.
It's very likely that this is an unintentional change. Please
double-check that all newly added files have a BSD-style license
header with the following copyright:
    Copyright 2013 The Flutter Authors. All rights reserved.
Files in 'third_party/txt' may have an Apache license header instead.
If you're absolutely sure that the change in license count is
intentional, update 'flutter/ci/licenses.sh' with the new count.
=================================================================
```

It seems like we should just say that files in third_party may have other licenses. Does that seem reasonable?

(Summarizing from discussion above, this is in third_party rather than buildroot/third_party because it's code generated from a web service; there's no upstream thing we could use to pull it via DEPS.)

**Comment by Hixie** [comment]
I'm confused by this. Third-party packages should come in via DEPS. Generated code should be generated at build time. Unless we are permanently forking this code the way we did with libtxt, it doesn't belong in flutter/engine.

**Comment by stuartmorgan** [comment]
> Generated code should be generated at build time.

The wrinkle is that it's generated by a web service.

But I just went and dug into this some more, and it looks like that's just a wrapper for https://github.com/Dav1dde/glad, which is command-line. Given that, we could presumably pull that repo via DEPS and then generate the files at built time using that python script.

**Comment by cbracken** [comment]
Assuming that the repo that @stuartmorgan points to generates the same code, that approach would definitely be preferable. @cloudwebrtc can you confirm that's the case?

**Comment by cloudwebrtc** [comment]
@cbracken @stuartmorgan @Hixie  Can we use python-pip in the buildroot environment? I just did the test and only need two commands to get the `glad` code we want, in (Windows/macOS/Ubuntu)

`pip install glad`
`glad --profile="compatibility" --api="gl=3.3" --generator="c" --spec="gl" --extensions="" --reproducible --out-path=glad_src_output`

While pulling the https://github.com/Dav1dde/glad repository to the local, you still need to run the following command to install the glad script.

`python setup.py install`

Run glad generate
`glad --profile="compatibility" --api="gl=3.3" ....`

**Comment by stuartmorgan** [comment]
Why do you want to install it? The readme had instructions for running it directly:

> Or launch glad directly (after cloning the repository):

**Comment by cloudwebrtc** [comment]
I added https://github.com/Dav1dde/glad to DEPS and checked out to `src/flutter/third_party/glad/scripts`, 

Then need to run

``python -m glad --profile="compatibility" --api="gl=3.3" --generator="c" --spec="gl" --extensions="" --reproducible --out-path=..``

in the `scripts` directory.

@cbracken Can you help me？

**Comment by stuartmorgan** [comment]
@cloudwebrtc It will be a week or two at least before I can take a look. If you want to experiment before then, what you want is an [`action`](https://gn.googlesource.com/gn/+/master/docs/reference.md#func_action) that has the generated files as outputs, and whatever it uses internally as inputs. Then targets that need the header would depend on that action target.

**Comment by cloudwebrtc** [comment]
@stuartmorgan Updated, please review again.

File: DEPS [hunk 8]
```diff
@@ -424,6 +424,9 @@ deps = {
   'src/third_party/pkg/when':
    Var('dart_git') + '/when.git' + '@' + '0.2.0',
 
+  'src/flutter/third_party/glad/scripts':
+   Var('github_git') + '/Dav1dde/glad.git' + '@' + 'de6c39e3040c987323b8ed078c36442f4fb681b3',
```
Comment (stuartmorgan) [inline DEPS]: @cbracken Do we want a mirror of this? It seems like we have other direct GitHub dependencies, but I'm not sure if there's a policy for when that's allowed.

File: third_party/glad/generate.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+# -*- coding: UTF-8 -*-
+
+import os
+import sys
+import argparse
+
+def generate(output):
+    dirname, _ = os.path.split(os.path.abspath(__file__))
+    command = 'cd ' + dirname + '/scripts;'
+    command += 'python -m glad --profile="compatibility" --api="gl=3.3" --generator="c" --spec="gl" --extensions="" --reproducible --quiet --out-path='
```
Comment (stuartmorgan) [inline third_party/glad/generate.py]: Is this really necessary? Why can't we run the glad script directly from GN? (Or at the very least, use the module directly in the wrapper rather that sub-invoking python from python.)

**Comment by cloudwebrtc** [comment]
I changed it to run `glad.__main__.py` directly in `generate.py`.

**Comment by stuartmorgan** [comment]
I can do a full re-review later, but from a quick look:

> and whatever it uses internally as inputs

This part is still missing. The `action` should have, as `inputs`, all the files that, were they to change, would require re-running this step. That means all the relevant py files, and from a brief glance it looks like there are some .xml files that are probably relevant as well. Without that, the step will never re-run if the outputs exist, even if `glad` is updated to a newer version, which will cause hard-to-find issues for some people who have stale generated files.

**Comment by cloudwebrtc** [comment]
> I can do a full re-review later, but from a quick look:
> 
> > and whatever it uses internally as inputs
> 
> This part is still missing. The `action` should have, as `inputs`, all the files that, were they to change, would require re-running this step. That means all the relevant py files, and from a brief glance it looks like there are some .xml files that are probably relevant as well. Without that, the step will never re-run if the outputs exist, even if `glad` is updated to a newer version, which will cause hard-to-find issues for some people who have stale generated files.

I don’t know much about how the action works, you mean [scripts/glad/files/gl.xml](https://github.com/Dav1dde/glad/blob/master/glad/files/gl.xml) as inputs to the action?
If we update the `Dav1dde/glad` repository, the action will detect the xml file, and if there is a change, the action will be re-executed.

**Comment by stuartmorgan** [comment]
An action runs when any of its inputs are newer than the outputs. For incremental build to work reliably you need to list **every** file that, were out to change, should cause the script to run again.

As a said above, that is all of the python files used by the script, as well as any data files.

**Comment by cloudwebrtc** [comment]
I need to use all the following files as inputs to the action. Is this correct?
`glad/*.py`
`glad/files/*.*`
`glad/lang/*.*` 

This is a list of files for the https://github.com/Dav1dde/glad/tree/master/glad repository.

**Comment by stuartmorgan** [comment]
You need to list every file that affects the output of the script. Whether or not that's every file in the repository I don't know; you'll need to look at the details of the script.

**Comment by cloudwebrtc** [comment]
Thanks, I can try to do it.

**Comment by cbracken** [comment]
@cloudwebrtc any luck with the new approach?

**Comment by cloudwebrtc** [comment]
@cbracken Sorry for the late update, @stuartmorgan I have added the relevant files to inputs and tested them. Modifying any .py file can trigger `action` again.

File: third_party/glad/BUILD.gn [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+  sources = [
+    "$_glad_gen_dir/src/glad.c",
+  ]
+
+  deps = [
+    ":gen_glad_sources",
+  ]
+}
+
+action("gen_glad_sources") {
```
Comment (stuartmorgan) [inline third_party/glad/BUILD.gn]: This rule should have file-private visibility.

File: third_party/glad/BUILD.gn [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    "scripts/glad/lang/c/loader/gl.py",
+    "scripts/glad/lang/c/loader/wgl.py",
+    "scripts/glad/lang/c/loader/egl.py",
+    "scripts/glad/lang/c/loader/glx.py",
+  ]
+
+  outputs = [
+    "$_glad_gen_dir/src/glad.c",
+    "$_glad_gen_dir/include/glad/glad.h",
+    "$_glad_gen_dir/include/KHR/khrplatform.h",
```
Comment (stuartmorgan) [inline third_party/glad/BUILD.gn]: Rather than duplicate these, please make file-local variables, one for the list of generated sources and one for the list of generated headers, and use those above and then again here (added together).

File: third_party/glad/BUILD.gn [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    "$_glad_gen_dir/src/glad.c",
+  ]
+
+  deps = [
+    ":gen_glad_sources",
+  ]
+}
+
+action("gen_glad_sources") {
+  script = "$flutter_root/third_party/glad/generate.py"
```
Comment (stuartmorgan) [inline third_party/glad/BUILD.gn]: Why is this not a relative path?

File: third_party/glad/BUILD.gn [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+  inputs = [
+    "generate.py",
```
Comment (stuartmorgan) [inline third_party/glad/BUILD.gn]: No need to list the script as an input; GN understands that if the script changes a rerun is required.

File: third_party/glad/generate.py [hunk 14]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/python
+# -*- coding: UTF-8 -*-
+
```
Comment (stuartmorgan) [inline third_party/glad/generate.py]: I'm still unclear on why we need this wrapper file. Please add a file-level comment here explaining why we need a wrapper script.

File: third_party/glad/LICENSE [hunk 15]
```diff
@@ -0,0 +1,25 @@
+Copyright 2013 The Flutter Authors. All rights reserved.
+
```
Comment (stuartmorgan) [inline third_party/glad/LICENSE]: Note that if we didn't need a wrapper script, we also wouldn't need this, and could just use src/third_party and put the GN file in the buildroot repo, just as with GLFW.

File: third_party/glad/BUILD.gn (hunk 10 above)
Comment (cloudwebrtc) [reply third_party/glad/BUILD.gn]: Sorry, I don't quite understand how to modify it.

File: third_party/glad/LICENSE (hunk 15 above)
Comment (cloudwebrtc) [reply third_party/glad/LICENSE]: I observed that GN can't run python modules in `python -m scripts/glad` mode, so generate.py seems to be necessary.

**Comment by cloudwebrtc** [comment]
@stuartmorgan Please review again.

File: third_party/glad/BUILD.gn (hunk 10 above)
Comment (stuartmorgan) [reply third_party/glad/BUILD.gn]: See the first example [here](https://gn.googlesource.com/gn/+/master/docs/reference.md#var_visibility)

File: third_party/glad/BUILD.gn [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    "scripts/glad/lang/c/generator.py",
+    "scripts/glad/lang/c/loader/__init__.py",
+    "scripts/glad/lang/c/loader/gl.py",
+    "scripts/glad/lang/c/loader/wgl.py",
+    "scripts/glad/lang/c/loader/egl.py",
+    "scripts/glad/lang/c/loader/glx.py",
+  ]
+
+  outputs = _public_headers
+  outputs += _glad_sources
```
Comment (stuartmorgan) [inline third_party/glad/BUILD.gn]: No need for two expressions here; you can just do `outputs = _public_headers + _glad_sources`.

**Comment by stuartmorgan** [comment]
Adding @cbracken as a reviewer for the question above.

**Comment by stuartmorgan** [comment]
Closing per https://github.com/flutter/engine/pull/9822#issuecomment-645668068. We can revisit if we end up needing GLFW texture support.
